[actions] fix "sh" broken when output is in encoding other than UTF-8

sdjmchattie opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue following this format: -->
Deals with logging of standard output from executed commands where the output is in encodings other than UTF-8.  Currently only UTF-16 and ISO-8859-1 are being handled and converted back to UTF-8, but these are the most common encodings.  Without this, the built in Ruby split command fails and raises an error.  This at least prevents most of these errors and if it makes a mistake with the encoding conversion then the worst that happens is that some less common characters are displayed incorrectly in the output.

Resolves #15153

### Description
<!-- Describe your changes in detail. -->
<!-- Please describe in detail how you tested your changes. -->
The change is fairly self explanatory, but basically we go through the possible encodings and force that encoding onto the string.  If Ruby then says the string has a valid encoding, we convert from that to UTF-8 for the split command to work.  The provided unit test checks that converting a string from UTF-16 and from ISO-8859-1 do not raise any errors now.  When the fix is removed, this test fails.

sdjmchattie committed:
Convert shell ASCII to UTF-8 before splitting strings

sdjmchattie committed:
Try common encodings until one provides valid encoding

sdjmchattie committed:
Rename method for encoding as utf 8

sdjmchattie committed:
Add rspec test to ensure alternatively encoded strings don't crash command_output

sdjmchattie commented:
@googlebot I signed it!

chilch reviewed (approved):
Looks good to me, thanks for fix!

ainame reviewed (commented):
It looks like your test case doesn't look right. The precondition of the scenario is not correct. (I've read through all #15153) What actually happens with the existing code for your issue's scenario is that the `String` object - `message` passed to `FastlaneCore::Shell#command_output` uses UTF-8 as its encoding while the real output has some UTF-16 characters, which is **invalid**. However, your test case prepared `String` objects encoded to UTF-16 or ISO-8859-1, which are both just **valid**.

For example, this runs successfully.

```ruby
p  "string_in_utf-16_with_µ\n".encode('UTF-16').split #=> ["\uFEFFstring_in_utf-16_with_\u00B5"]
```

I would suggest that you use `String#scrub` which eliminates/replaces invalid characters. This is super simple and does work for your scenario, I suppose. Most importantly this can cover all the possible encoding.

```ruby
actual = (message.scrub.split("\r").last || "") # as clearing the line will remove the `>` and the time stamp
```

A better test case being close to the issue's scenario would be like below.

```ruby
# Hard coded UTF-16 string but this String object is based on UTF-8
expect { shell.command_output("\xB5") }.not_to raise_error
```

What do you think?

sdjmchattie commented:
@ainame when I remove my fix from the code, the test fails. When I run the text file from #15153  through Fastlane (I.e. not in the unit test) it raises the error before my fix but works correctly after my fix. If you use `scrub` you would throw away good characters from the encoding used if you’re scrubbing while the string thinks it’s in UTF-8 encoding. With my fix, the command output contains all the correct characters in the output of the process. I think it’s fair to try to parse the string as UTF-16 if common tools like genstrings are likely to be parsed via Fastlane. It’s preferable to throwing characters away unnecessarily.

ainame commented:
@sdjmchattie 

> I think it’s fair to try to parse the string as UTF-16 if common tools like genstrings are likely to be parsed via Fastlane. 

I see your point. I missed the output you got were warnings from `genstrings`. Can you please share which encodings are necessary by which tools in your case? It's good to document that in code.

ainame commented on fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:
```diff
@@ -150,6 +150,19 @@ def password(message)
 
     private
 
+    def encode_as_utf_8_if_possible(message)
+      orig_encoding = message.encoding
+
+      # Try some common encodings and return the first with valid codings
+      ['UTF-8', 'UTF-16', 'ISO-8859-1'].each do |dest_encoding|
+        message.force_encoding(dest_encoding)
+        return message.encode('UTF-8', dest_encoding) if message.valid_encoding?
+      end
+
+      message.force_encoding(orig_encoding)
+      return message
+    end
```
And also as a piece of Ruby source code, this is not clear to me. I might want to improve this part.

This makes it a bit more Ruby-ish  code with using constant values and Object#dup not to mutate given parameter. Please document the use cases of encodings in the comment 🙇 

```suggestion
    def encode_as_utf_8_if_possible(message)
      return message if message.valid_encoding?
      
      # Try following common encodings and return the first with valid codings
      # 
      # * UTF-16 - used by genstrings
      # * ISO8859-1 - used by blah blah 
      test_string = message.dup
      [Encoding::UTF_16, Encoding::ISO8859_1].each do |dest_encoding|
        return message.encode(Encoding::UTF_8, dest_encoding) if test_string.force_encoding(dest_encoding).valid_encoding?
      end

      # return the original message if no valid encoding is found
      message
    end
```

ainame reviewed (changes requested):
Suggested some improvements in code.

sdjmchattie committed:
Fix some issues with unit test

sdjmchattie committed:
Use a more Rubyish encoding checker method

sdjmchattie commented:
Thanks @ainame I was a bit confused by your suggestions at first, but I've puzzled it out.  You are right that the string I generated in the unit test was wrong.  It was valid, as you suggested.  The problem that this is supposed to be resolving is that all standard output from other sources is assumed to be UTF-8 when it sometimes is not.  I've modified the unit test now to create a UTF-16 string (that is valid) and then invalidly force it to UTF-8 so that Ruby now thinks it's UTF-8.  This is the exact scenario that genstrings causes when we had the µ character in one of our strings.  I accepted your change to the make_utf8 method but also tweaked it to remove the ISO-8859-1 which I only included because it is a common encoding.  I don't know of a tool that outputs it specifically, but if you don't want it it wasn't causing a problem for me.

Hopefully the changes I've made will give you the "ahha" moment and that you can also now see that `p "hello\n".encode('UTF-16').force_encoding('UTF-8').split` raises an error.  The error I was seeing for the past 3 years with genstrings output.

sdjmchattie committed:
Merge remote-tracking branch 'upstream/master' into shell-stdout-fix-15153

ainame commented:
Can you please fix CI failures?

(Note that on the latest master branch I made a bug that CI checks pass when failing actually😅 but it will be resolved by this soon https://github.com/fastlane/fastlane/pull/18398 )

twocountriestrus commented:
Gracias

rogerluan commented:
Just adding my 2 cents:

I think this is very key in the issue:

>I realized that `NSLocalizedString("µ", comment: "")` reproduces a perfectly fine output, whereas if you add a letter `m` in front of it, it glitches, such as: `NSLocalizedString("µm", comment: "")`. Hm, interesting.

We should probably test that specific scenario to really make sure everything's 👌 😄

sdjmchattie committed:
Merge branch 'master' into shell-stdout-fix-15153

sdjmchattie committed:
Add an extra test for a string containing the µ character

minuscorp commented on fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:
```diff
@@ -150,6 +150,21 @@ def password(message)
+      #
+      # * UTF-16 - output by genstrings
+      test_string = message.dup
+      [Encoding::UTF_16].each do |dest_encoding|
```
Is this needed when there's only one encoding to be tested on?

minuscorp replied:
Yeah, that's why I asked of keeping the array now that we only have one left, not that it'll cost much computationally talking😂

sdjmchattie committed:
Merge branch 'master' into shell-stdout-fix-15153

sdjmchattie committed:
Remove array for alternative encodings

sdjmchattie commented:
I don't know how to fix the CI issues.  Some of these seem to change every time I merge master into this branch.
The previous run only failed on Ruby 2.4 but I can't reproduce the issue locally.
I installed Ruby 2.4.6p354 via `rbenv` so it's the exact same version as on the CI server and the test passes on my machine.

minuscorp commented:
Ruby 2.4 raises the next backtracked:
```
Failure/Error:
  expect do
    @shell.command_output("utf_16_string_tagged_as_utf_8\n".encode(Encoding::UTF_16).force_encoding(Encoding::UTF_8))
    @shell.command_output("µm\n".encode(Encoding::UTF_16).force_encoding(Encoding::UTF_8))
  end.not_to raise_error

  expected no Exception, got #<ArgumentError: invalid byte sequence in UTF-8>
```

sdjmchattie commented:
Sorry @minuscorp maybe I've missed the information you're giving me.  Here's what happens when I run this on my machine with the same version of Ruby:

![image](https://user-images.githubusercontent.com/1999241/112625077-5a4f9b00-8e26-11eb-992f-ffb3989c627b.png)

I don't know how to fix it if I can't see why it's broken.

minuscorp commented:
Maybe try to run the whole suite with a random order (see the command in the circleci) to see if it gets raised 🧐

ainame commented:
I may look into it later 🙂

Commit:
Fix for failing Ruby 2.4 test by removing all invalid characters

joshdholtz commented:
@googlebot I consent.

Commit:
Updated comment

joshdholtz commented:
@sdjmchattie @ainame I fixed the failing Ruby 2.4 test by replacing any invalid encoding with empty string 🤷‍♂️ Seemed like this was _only_ an issue in Ruby 2.4 for some unknown reason. Hope this is an okay fix 😇

sdjmchattie commented:
@googlebot I consent

sdjmchattie commented:
Thanks @joshdholtz I appreciate that. So weird that it had a problem on the CI server because locally I couldn't reproduce the issue.

ainame commented on fastlane_core/lib/fastlane_core/ui/implementations/shell.rb:
```diff
@@ -150,6 +150,17 @@ def password(message)
+      return message.encode(Encoding::UTF_8, Encoding::UTF_16) if test_message.force_encoding(Encoding::UTF_16).valid_encoding?
+
+      # replace any invalid with empty string
+      return message.encode(Encoding::UTF_8, invalid: :replace, replace: "")
```
Maybe it's worth leaving invalid characters as `�` as people would still want to if there were any chars there?


```suggestion
      return message.encode(Encoding::UTF_8, invalid: :replace)
```

https://ruby-doc.org/core-2.4.0/String.html#method-i-encode

ainame reviewed (commented):
I could confirm that it's failing on 2.4 on my end but @joshdholtz's fix looks good to me as there would be some histories in Ruby to dig around that but couldn't find it quickly🙂 I have a minor suggestion lastly!

Commit:
Update fastlane_core/lib/fastlane_core/ui/implementations/shell.rb

Co-authored-by: Satoshi Namai <s.namai.09@gmail.com>

joshdholtz reviewed (approved):
Thanks for fixing this, @sdjmchattie ❤️ Looks like this should be all good to go now. Great work, team!
